beamline/datautils.py

```diff
@@ -55,24 +59,80 @@ def getAllCols(self, sddsfile=None):
+
+        :param sddsfile: sdds file name, if not given, rollback to the one that from ``__init__()``
+        :return: all sdds data parameter names
+        :rtype: list
+
+        :Example:
+
+        >>> dh = DataExtracter('test.w1')
+        >>> print(dh.getAllPars())
+        ['Step', 'pCentral', 'Charge', 'Particles', 'IDSlotsPerBunch', 'SVNVersion', 'Pass', 'PassLength', 'PassCentralTime', 'ElapsedCoreTime', 'MemoryUsage', 's', 'Description', 'PreviousElementName']
+
+        :seealso: :func:`getAllCols`
+        """
+        if sddsfile is not None:
+            sddsobj = sdds.SDDS(2)
+            sddsobj.load(sddsfile)
+        else:
+            sddsobj = self.sddsobj
+        return sddsobj.parameterName
 
     def extractData(self):
-        """ extract data as numpy array, with given required fields,
-        put data into ``h5data`` as a numpy array.
+        """ return `self` with extracted data as `numpy array`
+
+        Extract the data of the columns and parameters of `self.kws` and put
+        them in a :np:func:`array` with all columns as columns or parameters as
+        columns. If columns and parameters are requested at the same then each column
+        is one row and all parameters are in the last row. This
+        :np:func:`array` is saved in ``h5data``.
+
+        .. note::
+            If you mix types (e. g. float and str) then the minimal fitting type is
+            taken for all columns.
+
+        :return: instance of itself
+
+        :Example:
 
-        :return: instance itself
+        One column of the watch element
+        >>> dh = DataExtracter('test.w1')
+        >>> dh.kwslist = ['Step']
+        >>> print(dh.extractData().h5data)
+        array([[1]])
+
+        Two columns of the watch element
+        >>> dh = DataExtracter('test.w1')
+        >>> dh.kwslist = ['s', 'betax']
+        >>> print(dh.extractData().h5data)
+        array([[0, 1], [1, 2], [2, 1]])
+
+        Two columns of the watch element and one parameter.
+        The columns transform to rows and the parameter row is at the end.
+        Furthermore all elements are strings, because the type of
+        `PreviousElementName` is str and not float.
+        >>> dh = DataExtracter('test.w1')
+        >>> dh.kwslist = ['s', 'PreviousElementName', 'betax']
+        >>> print(dh.extractData().h5data)
+        array([['0', '1', '2'], ['1', '2', '1'], ['DR01']])
         """
-        for k in self.kwslist:
-            self.dcmdline += ' -col={kw},format={p}'.format(kw=k, p=self.precision)
-        cmdlist = ['bash', self.dscript, self.dpath, self.dcmdline]
-        retlist = []
-        proc = subprocess.Popen(cmdlist, stdout=subprocess.PIPE)
-        for line in proc.stdout:
-            retlist.append([float(i) for i in line.split()])
-        self.h5data = np.array(retlist)
+        columns = self.sddsobj.columnName
+        parameters = self.sddsobj.parameterName
+        data = [self.sddsobj.columnData[columns.index(col)][0]
```

Comment:
In my laptop, `sddsobj.columnData` just gives a list of `[None]`, `sddsobj.parameterData` works right, so I cannot verify the test script here. Maybe it it the issue of the SDDS python interface.